### PR TITLE
Fix use_compute_types parameter name in CPP backend to_dtype

### LINKED ISSUE #167205: `torch.compile` produces different VJP gradients from eager mode in `float16`
### 🐛 Describe the bug

A gradient mismatch occurs between eager mode and `torch.compile` when using `torch.autograd.functional.vjp`. The discrepancy is observed when using `torch.float16` and appears to be dependent on the specific input values provided.

While the forward pass of the compiled function produces results that are numerically very close to the eager version, the backward pass (VJP) computes significantly different gradients for the inputs. This suggests a potential issue in how the backward pass is handled for one or more operations within the compiled graph at `float16` precision.

To reproduce:

```python
import torch

def robust_power(base: torch.Tensor, exponent: torch.Tensor, threshold: torch.Tensor) -> torch.Tensor:
    threshold1 = threshold
    result_shape = torch.broadcast_shapes(base.shape, exponent.shape)
    broadcasted_base = torch.broadcast_to(base, result_shape)
    broadcasted_exponent = torch.broadcast_to(exponent, result_shape)
    broadcasted_base = torch.abs(broadcasted_base)
    threshold = torch.broadcast_to(threshold.to(base.dtype), result_shape)
    cond = (broadcasted_exponent < 1 - torch.log(threshold) / torch.log(broadcasted_base)) & (broadcasted_base < threshold)
    output = torch.where(cond, 1 / threshold1 * broadcasted_base, broadcasted_base ** torch.where(cond, 1, exponent))
    return output


base = torch.tensor([[ 0.1331787109375,  0.071533203125, -0.154541015625, -0.0008382797241210938,
                       0.062103271484375, -0.072021484375,  0.0265655517578125,  0.0108489990234375,
                       0.0004291534423828125]], dtype=torch.float16)

exponent = torch.tensor([[ 0.1749267578125], [-0.0286102294921875]], dtype=torch.float16)
threshold = torch.tensor(6.103515625e-05, dtype=torch.float16)

out_grad = torch.tensor([
    [ 0.7158203125, -1.736328125,  1.9599609375, -1.8642578125, -1.720703125,
     -0.30810546875,  0.1309814453125, -1.3525390625, -0.429443359375],
    [-0.7373046875,  0.81591796875,  0.10162353515625,  0.062255859375,  0.587890625,
      0.1507568359375,  1.7099609375, -0.276611328125, -0.52978515625]
], dtype=torch.float16)

# ----------------------------
# Eager (uncompiled) execution
# ----------------------------
print("\n=== Eager Execution ===")
y_normal = robust_power(base, exponent, threshold)
print("Output:\n", y_normal)

y_n, grads_normal = torch.autograd.functional.vjp(
    lambda b, e, t: robust_power(b, e, t),
    (base, exponent, threshold),
    v=out_grad
)
grad_b_n, grad_e_n, _ = grads_normal
print("grad_b (normal):", grad_b_n)
print("grad_e (normal):", grad_e_n)


# ----------------------------
# Compiled execution
# ----------------------------
robust_power_compiled = torch.compile(robust_power)

print("\n=== Compiled function ===")
y_compiled = robust_power_compiled(base, exponent, threshold)
print("Output:\n", y_compiled)

y_c, grads_compiled = torch.autograd.functional.vjp(
    lambda b, e, t: robust_power_compiled(b, e, t),
    (base, exponent, threshold),
    v=out_grad
)
grad_b_c, grad_e_c, _ = grads_compiled
print("grad_b (compiled):", grad_b_c)
print("grad_e (compiled):", grad_e_c)


# ----------------------------
# Comparison
# ----------------------------
print("\n=== Comparison ===")
print("Output equal:", torch.allclose(y_normal, y_compiled, atol=1e-3, rtol=1e-3))
print("grad_b equal:", torch.allclose(grad_b_n, grad_b_c, atol=1e-3, rtol=1e-3))
print("grad_e equal:", torch.allclose(grad_e_n, grad_e_c, atol=1e-3, rtol=1e-3))

```
Results:
```
=== Eager Execution ===
Output:
 tensor([[0.7026, 0.6304, 0.7212, 0.2896, 0.6152, 0.6313, 0.5303, 0.4534, 0.2576],
        [1.0596, 1.0781, 1.0547, 1.2246, 1.0830, 1.0781, 1.1094, 1.1377, 1.2480]],
       dtype=torch.float16)
grad_b (normal): tensor([[  0.8291,  -3.0293,  -1.5811, 115.3750,  -3.2734,   0.5371,  -1.5840,
          -9.0625,  -1.1875]], dtype=torch.float16)
grad_e (normal): tensor([[ 9.8906],
        [-4.0156]], dtype=torch.float16)

=== Compiled function ===
Output:
 tensor([[0.7026, 0.6304, 0.7212, 0.2896, 0.6152, 0.6313, 0.5303, 0.4534, 0.2576],
        [1.0596, 1.0781, 1.0547, 1.2246, 1.0830, 1.0781, 1.1094, 1.1377, 1.2480]],
       dtype=torch.float16)
grad_b (compiled): tensor([[  0.8286,  -3.0273,  -1.5811, 115.3125,  -3.2734,   0.5371,  -1.5859,
          -9.0547,  -1.0000]], dtype=torch.float16)
grad_e (compiled): tensor([[ 9.8906],
        [-4.0117]], dtype=torch.float16)

=== Comparison ===
Output equal: True
grad_b equal: False
grad_e equal: True
```

### Versions

````
Collecting environment information...
PyTorch version: 2.9.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 15.6 (arm64)
GCC version: Could not collect
Clang version: 21.1.0
CMake version: version 4.0.1
Libc version: N/A

Python version: 3.12.8 (v3.12.8:2dc476bcb91, Dec  3 2024, 14:43:19) [Clang 13.0.0 (clang-1300.0.29.30)] (64-bit runtime)
Python platform: macOS-15.6-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M1 Pro

Versions of relevant libraries:
[pip3] Could not collect
[conda] numpy                     2.2.6                    pypi_0    pypi
[conda] torch                     2.7.0                    pypi_0    pypi
[conda] torchvision               0.22.0                   pypi_0    pypi


cc @frank-wei @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @chauhang @penguinwu

### PR DESCRIPTION
**Pull request by dsashidh**
Fixes #167205 

In reference to comment brought up by @eellison in the issue. 

Fixed parameter name mismatch in `CppVecOverrides.to_dtype()` that caused TypeError when enabling `emulate_precision_casts=True` on CPU. The parameter was named `use_compute_dtypes` but should be `use_compute_types` to match the base class and calling code in `lowering.py`. This prevented users from improving float16 gradient accuracy on CPU, a feature that already worked on CUDA.

Added regression test that verifies the config enables and computes VJP gradients without error. 

**NOTE**:  The test does not assert exact gradient equality as compilation naturally produces slightly different results due to operation reordering, vectorization, and platform-specific optimizations. This is expected behavior for compiled code. The fix significantly improves gradient accuracy while maintaining the performance benefits of compilation.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @chenyang78

### CODE REVIEW

File: test/inductor/test_cpu_repro.py [hunk 1]
```diff
@@ -4494,6 +4495,40 @@ def fn(x, weight, bias):
         torch.testing.assert_close(weight_cmp.grad, weight_ref.grad)
         torch.testing.assert_close(bias_cmp.grad, bias_ref.grad)
 
+    @config.patch(emulate_precision_casts=True)
+    def test_emulate_precision_casts_cpp_backend_no_error(self):
+        """
+        emulate_precision_casts threw TypeError on CPP backend.
```
Comment (jgong5) [inline test/inductor/test_cpu_repro.py:4518]: How about adding the link to the issue in the comment?
```suggestion
        See https://github.com/pytorch/pytorch/issues/167205
        emulate_precision_casts threw TypeError on CPP backend.
```

**Review by jgong5 — APPROVED** [review]
Thanks for the fix. LGTM.
